lib/Sema/CSGen.cpp

```diff
@@ -1727,9 +1727,11 @@ namespace {
       };
 
       // If a contextual type exists for this expression, apply it directly.
-      Optional<Type> arrayElementType;
-      if (contextualType &&
-          (arrayElementType = ConstraintSystem::isArrayType(contextualType))) {
+      if (contextualType && ConstraintSystem::isArrayType(contextualType)) {
+        contextualType = CS.replaceInferableTypesWithTypeVars(
```

Comment:
I'm a bit worried about the contextual type recorded in the constraint system not matching the one from constraint. 
This is similar situation to the one with opaque result types which need to be opened before use in constraints but recorded in their archetype for for AST coercion. I think a safer way might be to:
- Adjust `setContextualType` to "open" placeholder types and record both types in `contextualTypes` map
- Add a flag to `getContextualType` to indicate which of the types the caller wants.

This should help us remove a possibility of such bugs in the future.

Reply:
I initially tried doing the "opening" in `setContextualType` but the issue I was running into is that in situations where we _don't_ integrate the contextual type from the `contextualTypes` map into the constraint system, we end up with an orphaned type variable that never gets solved and so the expression is treated as "ambiguous".

E.g., with this setup:

```
struct S<T> {
    var x: T
}

var s: S<_> = .init(x: 0)
```

we bypass the `contextualTypes` map and integrate the `S<_>` type into the system via `generateInitPatternConstraints` which pulls the pattern type out from the initialization pattern via `ConstraintSystem::getTypeForPattern`, where the typed pattern handler calls `replaceInferableTypesWithTypeVars` itself.

So here, I wanted to wait to call `replaceInferableTypesWithTypeVars` until we were actually sure we were going to integrate the contextual type into the system, so that any type variables created would at least be hooked up to the system somehow...

Reply:
I such case let's open types when `getContextualType` requests "inner" type to be used in constraint, it could update `contextualTypes` map once the type is opened.

Reply:
Also, `generateInitPatternConstraints` shouldn't probably open contextual types directly anyway because that creates a discrepancy between what is stored in the constraint system and what is actually used.

Reply:
Hmm, but in the case of `visitArrayExpr` (and `visitDictionaryExpr`) we don't actually *use* the contextual type in the constraint system unless it actually is `Array` or `Dictionary`, so I think we would still end up with an orphaned type variable in the case where the contextual type is some other `ExpressibleByArrayLiteral` type with a placeholder. 

The goal of doing the opening here was to delay opening until we are 100% sure that we will use the opened type in the constraint.

(This wasn't an issue with unbound generic types because they fail the `isArrayType` and `isDictionaryType` checks so the unbound types never make their way into the constraint system.)

Reply:
I'm not sure I follow, we'd only one if something calls `getContextualType(<expr>, /*forConstraint=*/true)`, all other cases would produce an "outer" type which would still have placeholders in it.

Reply:
I guess another way to handle this would be to treat place holder is a regular type and add some logic to `matchTypes` to handle its opening when used in conversions so that it only is "opened" at the point of use if needs to be e.g. is there is a conversion from a concrete type like `Int` to placeholder type we just let that be simplified instead of doing anything. If there is a conversion between a type variable and placeholder we just don't attempt that as a viable binding.

Reply:
I actually think that this approach of handling placeholders while matching types and during binding inference might actually be more robust and future prove and should allow us to remove all of the `replaceInferableTypesWithTypeVars`.

Reply:
> we'd only one if something calls `getContextualType(<expr>, /*forConstraint=*/true)`, all other cases would produce an "outer" type which would still have placeholders in it.

I see, so the idea is that `visitArrayExpr` would first call `getContextualType (…, false)` to see if it had an array type, and only if we were going to actually make the constraint would it call `getContextualType(…, true)`?

> I actually think that this approach of handling placeholders while matching types and during binding inference might actually be more robust and future prove

Alright I’ll look into that. FWIW, most (all?) existing uses of `replaceInferable…` are places where we used to just do `openUnboundGenericType` which always did this sort of “up front” opening, so we’ll have to do the same thing for `UnboundGenericType`.

Reply:
> I see, so the idea is that `visitArrayExpr` would first call `getContextualType (…, false)` to see if it had an array type, and only if we were going to actually make the constraint would it call `getContextualType(…, true)`?

Yes, that's what I meant.